### ISSUE TITLE
Capacitor plugin

### ISSUE THREAD

**Issue by sandstrom** [OneSignal/OneSignal-Cordova-SDK#564]
Do you have plans to support Capacitor in OneSignal? 

It's a spiritual successor to Cordova, built by Ionic. With some adjustments this SDK should be fairly easy to adapt so that it'll work with Capacitor.

More details here:
https://capacitor.ionicframework.com/docs/plugins

**Comment by rgomezp** [comment]
Hello,
We don't currently plan to have an official OneSignal SDK for Capacitor. We always consider supporting more platforms if enough people request them. If you're interested, please show it by voting on this comment.

**Comment by henryavila** [comment]
Hi. I weally need the support for ionic capacitor. Any update on this? any plans to make OneSignal compatible com Capacitor?

**Comment by rgomezp** [comment]
Howdy y'all,
There has been significant interest in this issue as noted by the above comment upvotes.

Looking at the Capacitor documentation, our OneSignal-Cordova-SDK is most likely already compatible with Ionic Capacitor.

If your team is currently using Cordova and curious about moving to Capacitor, then it will be welcome to know that Capacitor has backwards compatible support for a large swath of existing Cordova plugins ([reference](https://ionicframework.com/resources/articles/capacitor-vs-cordova-modern-hybrid-app-development#h-migrating-to-capacitor-is-easy)).

Capacitor has support for most Cordova plugins, so developers can use the hundreds of existing Cordova plugins in their Capacitor apps ([reference](https://capacitorjs.com/docs/plugins/community#cordova-ecosystem)).

That said, and given the interest in this issue I can only assume that the integration in its current state doesn't work 100%. We have received reports of there being no issues as well as different reports that it doesn't work.

If anyone has more information regarding their integration process, please share it as context so that we can know how to proceed.

Please include things like:
- errors
- logs
- environment (version numbers etc...)

We appreciate all help from the community.

Cheers

**Comment by qmarcos** [comment]
Hi @rgomezp 

After reading this issue, I'm not sure if there are known problems integrating capacitor with onesignal currently. Is there any additional info on this since last comment?

Anyway I'm migrating an existing app from Phonegap Build to Capacitor right now so I think I could give you more visibility on this. Crossing fingers!

**Comment by jcardus** [comment]
I'm having issues, it doesn't give any error but no notification is shown. It get's delivered on signal dashboard

**Comment by rgomezp** [comment]
@qmarcos ,
Looking forward to your update.

@jcardus ,
Can you please include logs from the device with the OneSignal log level set to VERBOSE?

Cheers

**Comment by qmarcos** [comment]
@rgomezp after commenting on this issue I had to stop the migration (because other priorities)... so no updates in all this time. 

Probably next week I re-start the migration process so I hope to give you some feedback.

Thanks for the ping :)

**Comment by qmarcos** [comment]
Hi @rgomezp 

After some development I have reach the point to migrate the OneSignal Plugin from a Phonegap Build (PGB) based app to a new one based on Capacitor + Vuejs (developed from scratch, not based on Ionic). 

---

#### ⭕️ This is not a new app, so there are a some of certificates and configuration tasks that are already done. Others wre done for Capacitor following this docs guide: https://documentation.onesignal.com/docs/ionic-sdk-setup 

---

At this point I've tested on iOS only but it seems to work exactly the same way than when used with PGB... we have needed to double check all configurations options are ok on **xCode > App > Singing & Capabilities** (push and Background mode active, as is explained in the docs (https://documentation.onesignal.com/docs/ionic-sdk-setup)


And... we have nothing to change in our code base related to OneSignal Push notifications management!! 
All seems to work properly (tested in simulator and in real device). 

The code structure we are using is really simple:

### 1) Main initialisation function
```js
function initOneSignal() {
    // Enable to debug issues.    
    // window.plugins.OneSignal.setLogLevel({logLevel: 4, visualLevel: 4});

    // Callback triggered on notification opens
    var notificationOpenedCallback = function (jsonData) {
        // console.log('>>> notificationOpenedCallback: ' + JSON.stringify(jsonData));        

        // 👉 Your notification management code comes here
    };

    // Set iOS Settings
    var iosSettings = {};
    iosSettings["kOSSettingsKeyAutoPrompt"] = false;
    iosSettings["kOSSettingsKeyInAppLaunchURL"] = false;

    window.plugins.OneSignal
        .startInit( ONE_SIGNAL_APP_ID )
        .handleNotificationOpened(notificationOpenedCallback)
        .iOSSettings(iosSettings)
        .inFocusDisplaying(window.plugins.OneSignal.OSInFocusDisplayOption.None)
        .endInit();

    window.plugins.OneSignal.getIds(function (data) {            
        // At this point we set the device-token in our session and add it as a header for session persistance
        sessionStorage.setItem('DEVICE-TOKEN', data.userId);
        Vue.http.headers.common['Device-Token'] = data.userId;

        // 👉 Here is where we really launch our app        
    });
   
    // Request permissions 
    window.plugins.OneSignal.promptForPushNotificationsWithUserResponse(function(accepted) {
        console.log("User accepted notifications: " + accepted);
    });
};
```

### 2) Initialisation
We call to previous function in during the initialisation of the app (we have a small check to detect when the code is running under  an app enviroment or under a browser for base development):
```js
if (IS_APP) {
   // Other initialisation stuff
    initOneSignal();
} else {
// Other initialisation stuff for browser enviroment
}
```


Hope it helps, if I found problems testing against android I'll post them here :)

**Comment by rgomezp** [comment]
@qmarcos ,
Thank you very much for your detailed post.

It seems that the main thing is merely confusion around whether OneSignal supports Capacitor in the first place. We will make this more clear in the documentation.

TLDR; Capacitor is supported!

For that reason, I will go ahead and close this issue. If there are any other problems, please open a new issue.

Enjoy y'all

**Comment by sandstrom** [comment]
@rgomezp A few ideas on things you can improve in the docs (https://documentation.onesignal.com/docs/ionic-sdk-setup):

- Is `OneSignalNotificationServiceExtension` required, or optional? Says its only needed for rich text. What happens if I skip that step?
- Spelling mistake in "(For Kindle FIre Apps)"
- The block "Add required code" has no example for Capacitor (some other code blocks do).
- The "Update the OneSignal SDK" code block has no example for Capacitor (some other blocks do)
- Requirements is somewhat ambiguous on requirements, e.g. it says "Cordova 8.0.0 or newer" but I'm guessing you'd want to write "Cordova >= 8.0 or Capacitor >= 2.0" 

In this document I'd also modify a few things (https://documentation.onesignal.com/docs/troubleshooting-cordova-variants):

- "You will want to change to PhoneGap CLI to continue using the OneSignal SDK." change it to "You will want to change to PhoneGap CLI (or switch to Cordova or Capacitor) to continue using the OneSignal SDK."
- "Use the latest Cordova Version" change to "Use the latest Cordova or Capacitor Version"
- "Update OneSignal" needs example for Capacitor (use the code blocks with multiple script types, as you have in other places.

**Comment by rgomezp** [comment]
That is super helpful @sandstrom . We very much appreciate your input.

Cheers
